### PR TITLE
TriggerAuth-podIdentity.identityId - validation removed (operator)

### PR DESCRIPTION
**Pull request by radekfojtik**
TriggerAuthentication.podIdentity - validation of `identityId` (empty string "") removed (operator)
- validation remains only in webhook

### Checklist

- [x] When introducing a new scaler, I agree with the [scaling governance policy](https://github.com/kedacore/governance/blob/main/SCALERS.md)
- [x] I have verified that my change is according to the [deprecations & breaking changes policy](https://github.com/kedacore/governance/blob/main/DEPRECATIONS.md)
- [x] Tests have been added
- [x] Changelog has been updated and is aligned with our [changelog requirements](https://github.com/kedacore/keda/blob/main/CONTRIBUTING.md#Changelog)
- [x] A PR is opened to update our Helm chart ([repo](https://github.com/kedacore/charts)) *(if applicable, ie. when deployment manifests are modified)*
- [x] A PR is opened to update the documentation on ([repo](https://github.com/kedacore/keda-docs)) *(if applicable)*
- [x] Commits are signed with Developer Certificate of Origin (DCO - [learn more](https://github.com/kedacore/keda/blob/main/CONTRIBUTING.md#developer-certificate-of-origin-signing-your-work))

Fixes https://github.com/kedacore/keda/issues/5109

Relates to #

### CODE REVIEW

**Comment by radekfojtik** [comment]
@JorTurFer @tomkerkhove please take a look guys

File: apis/keda/v1alpha1/triggerauthentication_webhook.go [hunk 1]
```diff
@@ -107,15 +107,24 @@ func isTriggerAuthenticationRemovingFinalizer(om metav1.ObjectMeta, oldOm metav1
...
 		default:
-			return nil, nil
+			return nil
 		}
 	}
-	return nil, nil
+	return nil
+}
+
+func isEmptyString(str *string) bool {
```
Comment (josefkarasek) [inline apis/keda/v1alpha1/triggerauthentication_webhook.go:128]: Util functions like this one have a high potential for code duplicity. Can you add a new file to `pkg/util` and move the function there? Ideally with generics, something like

```golang
func IsDefaultValue[T comparable](value *T) bool {
	var def T
	return value != nil && *value == def
}
```

File: apis/keda/v1alpha1/triggerauthentication_webhook.go [hunk 2]
```diff
@@ -107,15 +107,24 @@ func isTriggerAuthenticationRemovingFinalizer(om metav1.ObjectMeta, oldOm metav1
...
+}
+
+func validatePodIdentityID(podIdentity *AuthPodIdentity) error {
+	if podIdentity != nil {
+		switch podIdentity.Provider {
 		case PodIdentityProviderAzure, PodIdentityProviderAzureWorkload:
-			if spec.PodIdentity.IdentityID != nil && *spec.PodIdentity.IdentityID == "" {
-				return nil, fmt.Errorf("identityid of PodIdentity should not be empty. If it's set, identityId has to be different than \"\"")
+			if isEmptyString(podIdentity.IdentityID) {
+				return fmt.Errorf("identityid of PodIdentity should not be empty. If it's set, identityId has to be different than \"\"")
```
Comment (josefkarasek) [inline apis/keda/v1alpha1/triggerauthentication_webhook.go:119]:
```suggestion
				return fmt.Errorf("identityId of PodIdentity should not be empty. If it's set, identityId has to be different than \"\"")
```

**Review by josefkarasek — APPROVED** [review]
I tested the PR locally with:

1. Install keda v2.11.0 with helm
2. Create `TriggerAuthentication` with `identityId: ""`
3. Upgrade to keda v.2.12.0 - I was able to reproduce the issue
4. Deploy patched keda and validation webhook images, include `ValidatingWebhookConfiguration` from charts:main branch

Old `TriggerAuthentication`s with `identityId: ""` don't cause issues and the validation webhook blocks the creation of new ones.

**Comment by tomkerkhove** [comment]
100%

**Comment by JorTurFer** [comment]
> Just a side note, we shouldn't rely on validation webhooks that much. If possible, this should work withouth validation as well.

I have to disagree with this point. KEDA has to work properly handling the scenario, but the data validation is the responsibility of the webhooks, don't you think? I mean, why do we use validating webhooks for validating the data plane, if we are going to raise validation messages in the operator too? 

KEDA operator will fail with a handled error if the user is providing the item wrongly (it won't panic), so other validations over the data constraints should be managed by the webhooks and users have to assume that validations are done by webhooks. As user, you can't skip the validating webhooks and also expect a well-formed validation with a lot of info about the validating error

**Comment by JorTurFer** [comment]
LGTM but let's wait until @zroubalik and @tomkerkhove agree with it, because this moves the responsibility from the operator to webhooks

**Comment by tomkerkhove** [comment]
> I have to disagree with this point. KEDA has to work properly handling the scenario, but the data validation is the responsibility of the webhooks, don't you think?
> [...]
>  As user, you can't skip the validating webhooks and also expect a well-formed validation with a lot of info about the validating error

No, because webhooks are optional. If it would be mandatory then I'd agree.

Validation webhooks are more for enforcing best practices in the current state of KEDA, not for ensuring things are valid. If we want to go that route than making it mandatory is the only viable option.

**Comment by radekfojtik** [comment]
@zroubalik @tomkerkhove Having this validation on the `operator` side is a breaking change introduced in v2.12 (is not backwards-compatible). I would say it needs to be removed now anyway. Alternatively, it is possible to add a warning and introduce the validation in a later version. But interesting - the operator will warn the user about a non-valid value that was not added by the user but by default by the operator.

**Comment by radekfojtik** [comment]
> I tested the PR locally with:
> 
> 1. Install keda v2.11.0 with helm
> 2. Create `TriggerAuthentication` with `identityId: ""`
> 3. Upgrade to keda v.2.12.0 - I was able to reproduce the issue
> 4. Deploy patched keda and validation webhook images, include `ValidatingWebhookConfiguration` from charts:main branch
> 
> Old `TriggerAuthentication`s with `identityId: ""` don't cause issues and the validation webhook blocks the creation of new ones.

@josefkarasek There is no need to set the `identityId: ""`, the `operator` does it for you in case you omit this property (when the finelizer is added). Thanks for suggestions.

**Comment by JorTurFer** [comment]
> @JorTurFer could you please look at this? We should resolve this already :) Thanks

I reviewed here: https://github.com/kedacore/keda/pull/5142#issuecomment-1821669062 

From my perspective, webhooks have to validate the information and the operator has to just handle errors if they are. No having VALIDATING webhooks IMHO explicitly means that you don't have validations (because you are not installing them), just fails if you introduce something wrong. 
In my opinion, this removal is 100% correct:
image

It moves the responsibility from the operator to the webhooks. If the operator sees and empty pod identity, it doesn't matter who sets it, and it continues with its work, failing if it has to fail

**Comment by JorTurFer** [comment]
We should push this topic.

**Comment by zroubalik** [comment]
@JorTurFer let's add it to the agenda for the next meeting

**Comment by tomkerkhove** [comment]
Sorry, I did not see the pings - Let's discuss indeed

File: apis/keda/v1alpha1/triggerauthentication_webhook.go (hunk 1 above)
Comment (JorTurFer) [reply apis/keda/v1alpha1/triggerauthentication_webhook.go:128]: I think that this is a good feedback ❤️

**Pull request closed without merging**